kernel/timeout.c

```diff
@@ -68,8 +68,14 @@ static int32_t next_timeout(void)
 {
 	struct _timeout *to = first();
 	int32_t ticks_elapsed = elapsed();
-	int32_t ret = to == NULL ? MAX_WAIT
-		: CLAMP(to->dticks - ticks_elapsed, 0, MAX_WAIT);
+	int32_t ret;
+
+	if ((to == NULL) ||
+	    ((int64_t)(to->dticks - ticks_elapsed) > (int64_t)INT_MAX)) {
```

Comment:
You want the cast interior to this subtraction.  If CONFIG_TIMEOUT_64BIT=n then the result will be computed in 32 bit precision and thus the comparison will be always false, which is not what you want.

Also IIRC the bug being addressed here is that to->dticks may be equal to K_TICKS_FOREVER (== -1).   Won't this still break?  Imagine ticks_elapsed == 2, so dticks of FOREVER will be -1 - 2 == "-3", which is less than INT_MAX so it will fail to clamp to MAX_WAIT correctly.

But then, it seems like another comment in the bug claims that in fact dticks is **never** set to K_TICKS_FOREVER (which might be true!  Refactoring elsewhere might have left dead code here).  In which case I'm confused about why we need the change at all.

Reply:
the problem is that MAX_WAIT is `K_TICKS_FOREVER` when this option is set. In this case
```
CLAMP(to->dticks - ticks_elapsed, 0, MAX_WAIT)
```
MAX_WAIT will be < to->dticks - ticks_elapsed and consequently the returned value will always be `K_TICKS_FOREVER`.

Reply:
just to be clear the bug is *NOT* for `to->dticks == K_TICKS_FOREVER`

Reply:
Ah, that's right.  Got it.  Still needs the precision fix though: as written with 32 bit timeouts it's impossible to compute dticks-elapsed to be >INT_MAX.